overlord/servicestate/quota_handlers.go

```diff
@@ -781,31 +802,41 @@ func ensureSnapServicesStateForGroup(st *state.State, grp *quota.Group, opts *en
 	return restartSnapServices(st, appsToRestartBySnap)
 }
 
-func ensureGroupIsNotMixed(group string, allGrps map[string]*quota.Group) error {
+// ensureGroupIsNotMixed is a helper which returns an error if the snaps and/or
+// services we are trying to add would result in a mixed group. A group is considered
+// mixed if:
+// 1. The group would end up containing both snaps and services
+// 2. The group would end up containing both services and groups
+func ensureGroupIsNotMixed(snapsToAdd, servicesToAdd []string, group string, allGrps map[string]*quota.Group) error {
+	if len(snapsToAdd) > 0 && len(servicesToAdd) > 0 {
+		return fmt.Errorf("cannot mix services and snaps in the same quota group")
+	}
+
 	grp, ok := allGrps[group]
-	if ok && len(grp.SubGroups) != 0 && len(grp.Snaps) != 0 {
-		return fmt.Errorf("quota group %q has mixed snaps and sub-groups, which is no longer supported; removal and re-creation is necessary to modify it", group)
+	if ok && ((len(grp.Services) > 0 && len(snapsToAdd) > 0) || (len(grp.Snaps) > 0 && len(servicesToAdd) > 0)) {
+		return fmt.Errorf("cannot mix services and snaps in the same quota group")
 	}
 	return nil
 }
 
-// ensureGroupHasNoSubgroups returns true if the group has no sub-groups or it doesn't exist,
-// otherwise turns false.
-func ensureGroupHasNoSubgroups(group string, allGrps map[string]*quota.Group) bool {
-	grp, ok := allGrps[group]
-	if ok && len(grp.SubGroups) != 0 {
-		return false
+// ensureParentIsNotMixed returns true if the parent is not a mixed group. A group
+// is considered mixed if it contains both snaps and sub-groups. In this case we only
+// allow services to be added to sub-groups, as we do not support nesting of snaps.
+func ensureParentIsNotMixed(parentGroup string, allGrps map[string]*quota.Group) bool {
+	parent := allGrps[parentGroup]
+	if parent == nil {
+		return true
 	}
-	return true
+	return len(parent.SubGroups) == 0 || len(parent.Snaps) == 0
 }
 
-func validateSnapForAddingToGroup(st *state.State, snaps []string, group string, allGrps map[string]*quota.Group) error {
+func validateSnapForAddingToGroup(st *state.State, snaps []string, group, parentGroup string, allGrps map[string]*quota.Group) error {
```

Comment:
I would pass the parentGroup as a Group or nil, this avoids the order confusion between two strings